test/object-store/c_api/c_api.cpp

```diff
@@ -3962,6 +3962,163 @@ struct Userdata {
+    };
+
+    realm_sync_config_t* local_sync_config = static_cast<realm_sync_config_t*>(local_config.sync_config.get());
+
+    SECTION("Manual reset") {
+        realm_sync_config_set_resync_mode(local_sync_config, RLM_SYNC_SESSION_RESYNC_MODE_MANUAL);
+
+        struct ResetRealmFiles {
+            void set_app(std::shared_ptr<realm::app::App> app)
+            {
+                m_app = app;
+            }
+            void do_reset(const char* path)
+            {
+                realm_app_t realm_app{m_app};
+                realm_sync_immediately_run_file_actions(&realm_app, path);
+            }
+            static ResetRealmFiles& instance()
+            {
+                static ResetRealmFiles instance;
+                return instance;
+            }
+            std::shared_ptr<realm::app::App> m_app;
+        };
+
+        ResetRealmFiles::instance().set_app(app);
+        realm_sync_config_set_error_handler(
+            local_sync_config,
+            [](realm_userdata_t, realm_sync_session_t*, const realm_sync_error_t sync_error) {
+                REQUIRE(sync_error.c_original_file_path_key);
+                REQUIRE(sync_error.c_recovery_file_path_key);
+                REQUIRE(sync_error.is_client_reset_requested);
+                ResetRealmFiles::instance().do_reset(sync_error.c_original_file_path_key);
+                ClientSyncObject::instance().signal();
+            },
+            nullptr, nullptr);
+
+        make_reset(local_config, remote_config)
+            ->on_post_reset([&](SharedRealm) {
+                util::EventLoop::main().run_until([&] {
+                    return ClientSyncObject::instance().stop();
+                });
+            })
+            ->run();
+    }
+
+    SECTION("Local Discard") {
+        realm_sync_config_set_resync_mode(local_sync_config, RLM_SYNC_SESSION_RESYNC_MODE_DISCARD_LOCAL);
+
+        struct AtomicCounter {
+            void increment()
+            {
+                std::lock_guard<std::mutex> lk{m_mutex};
+                m_counter += 1;
+            }
+            std::size_t fetch()
+            {
+                std::lock_guard<std::mutex> lk{m_mutex};
+                return m_counter;
+            }
+            std::mutex m_mutex;
+            std::size_t m_counter{0};
+        };
+
+        struct BeforeRst : AtomicCounter {
+            static BeforeRst& instance()
+            {
+                static BeforeRst instance;
+                return instance;
+            }
+        };
+
+        struct AfterRst : AtomicCounter {
+            static BeforeRst& instance()
+            {
+                static BeforeRst instance;
+                return instance;
+            }
+        };
+
+        realm_sync_config_set_before_client_reset_handler(
+            local_sync_config,
+            [](realm_userdata_t, realm_t*) -> bool {
+                BeforeRst::instance().increment();
+                return true;
```

Comment:
This is a good test of the happy path. Do we want to do the same test again but return false in the before handler to make sure that the after handler is not called, and that the error handler is called instead? IIRC that was the use case that created the discussion about testing these APIs.